### ISSUE TITLE
[Cross-posted from the Forum/Suggestion] Implement a way to integrate (original image file, detected text) →searchable PDF

### ISSUE THREAD

**Issue by Wikinaut** [tesseract-ocr/tesseract#660]
https://groups.google.com/forum/#!topic/tesseract-ocr/vvMldrkcuOQ has asked:
> I have a pdf (scanned) and now i make a searchable pdf from this.
>First i generate a black/white multipage tif, and with tesseract i can make a searchable pdf.
> But is it somehow possible to integrate the original pdf images?
> because the generated tif has not the same quality like the original (maybe the scaned image is in color).

How to reproduce:

1. Assume one page with a colored background `in.pdf`, converted to `in.ppm` image
1. preprocess `unpaper in.ppm in-cleaned.ppm` 
1. process with (example) `tesseract in-cleaned.ppm out -l deu+eng --oem 2 pdf txt`
1. tesseract mixed output file `out.pdf`has now a blotchy background (from the `unpaper` step above)

![20170113-10 09 17_auswahl](https://cloud.githubusercontent.com/assets/1151915/21924224/72ad2b02-d978-11e6-8b44-12c03da58aa2.png)

Is there any way to "feed-in" the original `in.ppm` as image, so that this is used instead of `in-cleaned.ppm` when creating the `out.pdf` ?

So what is wanted is original input image plus ocr layer, so that output looks like
![20170113-10 12 22_auswahl](https://cloud.githubusercontent.com/assets/1151915/21924289/db580474-d978-11e6-8fd6-05670345ff34.png)

**Comment by jbarlow83** [comment]
This is a complicated way of asking for an option to send one image through OCR and insert a different image in the output PDF.

```
tesseract --pdf-image original.png cleaned.png -l eng --oem 2 pdf  # not implemented, could work like this
```

I know this was requested before and I believe @jbreiden said it would be added to the PDF renderer  at some point.

**Comment by jbreiden** [comment]
I'm very reluctant to make Tesseract PDF generation fancy. I wonder if we can do an image swap like this outside of Tesseract, using one of the PDF manipulation toolkits.

**Comment by jbarlow83** [comment]
Sounds reasonable.

It is fairly simple to swap an image using qpdf's C++ API.

**Comment by Wikinaut** [comment]
@jbreiden It's the last really missing issue. 
The new algorithm is already a boost in quality,

If this helps, I will donate some mBTC for implementing it just right now. Just post your receiving address.

**Comment by Wikinaut** [comment]
@jbarlow83 background info. As you know, I recently wanted to try your OCRmyPDF because I found the interesting `-clean` option (source: https://media.readthedocs.org/pdf/ocrmypdf/latest/ocrmypdf.pdf ) which would have solved my problem:

which "does not alter the final output":

```
--clean
uses unpaper to clean up pages before OCR, but does not alter the final output.  This makes it less
likely that OCR will try to find text in background noise.
•
--clean-final
uses unpaper to clean up pages before OCR and inserts the page into the final output.  You
will want to review each page to ensure that unpaper did not remove something important.
```
but unfortunately this does not work with tesseract 4, at the present.

So I looked for bug reports, if tesseract could pass the original input image to the output; and filed the present issue.

**Comment by jbreiden** [comment]
Really? That's interesting, qpdf is very well written. Maybe the right thing to do is allow Tesseract to produce a multi-page PDF with invisible symbolic text PDF only, no images. Then another tool (perhaps an enhanced qpdf tool) would merge and composite two PDFs together. One being the original image-only PDF, and the other an invisible-text-only  PDF. What do you think, @jbarlow83? Please point me at the relevant qpdf API calls if you happen to know them.

**Comment by jbarlow83** [comment]
I think invisible text only output would be far more useful for developers that integrate tesseract or anyone who wants to do something fancy. It would still make sense to keep the existing OCR with image option of course. As a plus, it's should be easier to suppress the image than add a different one.

OCRmyPDF (which I maintain) use Ghostscript to rasterize and then runs one of its two PDF renderers. One uses Tesseract hOCR and provides more features but is not as good at producing the OCR text layer as Tesseract PDF, so I also provide Tesseract PDF. If Tesseract could produce a invisible text only I could offer all the features for both, and work towards phasing out the hOCR renderer. When possible I already do graft the text layer onto the existing PDF instead of constructing a new one.

In addition to OCRmyPDF `pdftk multibackground` could merge an OCR layer onto an existing PDF (by "watermarking"), so there is at least one other supported tool out there that should work out of the box. There's some other tools that wrap tesseract for use with PDFs as well.

In writing this I've made a case for not using qpdf because other tools should be able to do the job with an invisible text PDF, but for interest's sake case here is [example code](
https://github.com/qpdf/qpdf/blob/master/examples/pdf-invert-images.cc) that inverts black and white for all images; clearly this is close to how one would replace an image outright.

**Comment by jbreiden** [comment]
This sounds reasonable to me. I'll try to find time over this coming week to make an experimental invisible-text-only PDF that we can play with. All the other pieces of the puzzle are there; for example Leptonica already ships with a images->pdf tool that avoids transcoding for PNG, JP2K, and JPEG. It would be cool to use qpdf for the merge step because it is already so useful for linearizing. But it's great that there are more options. The qpdf author is extremely friendly in my experience, in case we eventually chat with him. Oh, I now vaguely remember that PDFBox had something for merging as well, but I've never tried it and can't find it at the moment.

**Comment by jbreiden** [comment]
Here's an experimental PDF pair, image-only and text-only. Let the merging begin!

[images.pdf](https://github.com/tesseract-ocr/tesseract/files/712576/images.pdf)
[text.pdf](https://github.com/tesseract-ocr/tesseract/files/712628/text.pdf)

**Comment by jbreiden** [comment]
This works brilliantly. I will implement for real if someone promises that they will use it. Also, what do we call the configuration option? My best idea so far to describe a PDF that has invisible text only is 'naked'. I'm sure someone has a better idea. 

```
$ time pdftk text.pdf multibackground images.pdf output full.pdf
real	0m0.253s
```

**Comment by jbreiden** [comment]
Implementation complete and under review by Ray.

**Comment by jbarlow83** [comment]
Looks really good @jbreiden. 

Works great in pdftk. No display issues and PDF syntax looks fine.

PyPDF2 is also capable of merging. It does not have the equivalent of "multibackground" but merge pages manually. Here is merging one page:

```python
In [1]: import PyPDF2 as pypdf

In [4]: pdf_text = pypdf.PdfFileReader(open('text.pdf', 'rb'))

In [5]: pdf_image = pypdf.PdfFileReader(open('images.pdf', 'rb'))

In [6]: page_text = pdf_text.pages[1]

In [7]: page_image = pdf_image.pages[1]

In [8]: page_text.mergeRotatedScaledTranslatedPage(page_image, 0, 1.0, 0, 0, expand=False)

In [9]: out = pypdf.PdfFileWriter()

In [10]: out.addPage(page_text)

In [11]: with open('pypdfmerge.pdf','wb') as o:
    ...:     out.write(o)
    ...:     

```

For reference, pdfbox did not work out of the box. As far as I can tell the closest command in pdfbox is
```
java -jar pdfbox-app-2.0.2.jar OverlayPDF images.pdf text.pdf pdfboxoverlay.pdf
```

However pdfbox takes the unusual approach of rasterizing the overlay PDF as a bitmap and drawing it on top of the base page, making it useless regardless of image/text order. (I suppose when you go to the trouble implementing a full PDF renderer in Java you feel compelled to use it even when it's not strictly needed.)

**Comment by jbarlow83** [comment]
I don't know about calling it a naked PDF because there's nothing exciting to see in it. It's more of a phantom or spectral apparition PDF, having form without substance.

`ocr_text_only` would do, or `suppress_images`? Not nearly as fun....

**Comment by jbreiden** [comment]
Spectral writing. Perhaps a kind of ghost script, if you will.

**Comment by Shreeshrii** [comment]
How about `text_only_pdf` ?

@jbreiden  is it also possible to use a .pdf file as input to tesseract directly?

**Comment by amitdo** [comment]
`pdf_invisible_text_layer_only`
+
a config file`pdfinvisible`

**Comment by jbarlow83** [comment]
@Shreeshrii PDF is a very complex vector-based file format. Tesseract works only on images. It is much easier to write PDFs that use a limited set of PDF features than read arbitrary PDFs. Have a look at OCRmyPDF (which I develop) - it addresses the details of using tesseract to apply OCR to PDFs.

**Comment by Wikinaut** [comment]
@jbreiden @jbarlow83 @amitdo  info: I just built the *whole*  toolchain from their git repos (tesseract, ocrmypdf, unpaper), and have ghostscript version 9.20 ready in a dedicated debian 9 "OCR VM" on my Qubes OS system.

Pls. let me know, what (if) you want me to test - I have time to test and want to help you.

**Comment by jbreiden** [comment]
Hmmm, an invisible text layer, invisible text, let's see ...  iText? Anyway, I'll pick something. There is zero chance that a PDF rasterizer will ever be part of Tesseract or Leptonica. In theory one could write an PDF image extractor for Leptonica, but there isn't really enough motivation to do so.

**Comment by jbreiden** [comment]
```diff
--- api/pdfrenderer.cpp	2016-12-13 14:43:24.000000000 -0800
+++ tesseract/api/pdfrenderer.cpp	2017-01-19 14:50:56.000000000 -0800
@@ -178,10 +178,12 @@
  * PDF Renderer interface implementation
  **********************************************************************/
 
-TessPDFRenderer::TessPDFRenderer(const char* outputbase, const char *datadir)
+TessPDFRenderer::TessPDFRenderer(const char *outputbase, const char *datadir,
+                                 bool textonly)
     : TessResultRenderer(outputbase, "pdf") {
   obj_  = 0;
   datadir_ = datadir;
+  textonly_ = textonly;
   offsets_.push_back(0);
 }
 
@@ -326,7 +328,11 @@
   pdf_str.add_str_double("", prec(width));
   pdf_str += " 0 0 ";
   pdf_str.add_str_double("", prec(height));
-  pdf_str += " 0 0 cm /Im1 Do Q\n";
+  pdf_str += " 0 0 cm";
+  if (!textonly_) {
+    pdf_str += " /Im1 Do";
+  }
+  pdf_str += " Q\n";
 
   int line_x1 = 0;
   int line_y1 = 0;
@@ -832,6 +838,7 @@
 bool TessPDFRenderer::AddImageHandler(TessBaseAPI* api) {
   size_t n;
   char buf[kBasicBufSize];
+  char buf2[kBasicBufSize];
   Pix *pix = api->GetInputImage();
   char *filename = (char *)api->GetInputName();
   int ppi = api->GetSourceYResolution();
@@ -840,6 +847,9 @@
   double width = pixGetWidth(pix) * 72.0 / ppi;
   double height = pixGetHeight(pix) * 72.0 / ppi;
 
+  snprintf(buf2, sizeof(buf2), "XObject << /Im1 %ld 0 R >>\n", obj_ + 2);
+  const char *xobject = (textonly_) ? "" : buf2;
+
   // PAGE
   n = snprintf(buf, sizeof(buf),
                "%ld 0 obj\n"
@@ -850,19 +860,18 @@
                "  /Contents %ld 0 R\n"
                "  /Resources\n"
                "  <<\n"
-               "    /XObject << /Im1 %ld 0 R >>\n"
+               "    %s"
                "    /ProcSet [ /PDF /Text /ImageB /ImageI /ImageC ]\n"
                "    /Font << /f-0-0 %ld 0 R >>\n"
                "  >>\n"
                ">>\n"
                "endobj\n",
                obj_,
-               2L,            // Pages object
-               width,
-               height,
-               obj_ + 1,      // Contents object
-               obj_ + 2,      // Image object
-               3L);           // Type0 Font
+               2L,  // Pages object
+               width, height,
+               obj_ + 1,  // Contents object
+               xobject,   // Image object
+               3L);       // Type0 Font
   if (n >= sizeof(buf)) return false;
   pages_.push_back(obj_);
   AppendPDFObject(buf);
@@ -899,13 +908,15 @@
   objsize += strlen(b2);
   AppendPDFObjectDIY(objsize);
 
-  char *pdf_object;
-  if (!imageToPDFObj(pix, filename, obj_, &pdf_object, &objsize)) {
-    return false;
+  if (!textonly_) {
+    char *pdf_object = nullptr;
+    if (!imageToPDFObj(pix, filename, obj_, &pdf_object, &objsize)) {
+      return false;
+    }
+    AppendData(pdf_object, objsize);
+    AppendPDFObjectDIY(objsize);
+    delete[] pdf_object;
   }
-  AppendData(pdf_object, objsize);
-  AppendPDFObjectDIY(objsize);
-  delete[] pdf_object;
   return true;
 }
 

--- api/renderer.h	2016-11-07 07:44:03.000000000 -0800
+++ api/renderer.h	2017-01-19 14:50:56.000000000 -0800
@@ -186,7 +186,7 @@
  public:
   // datadir is the location of the TESSDATA. We need it because
   // we load a custom PDF font from this location.
-  TessPDFRenderer(const char *outputbase, const char *datadir);
+  TessPDFRenderer(const char* outputbase, const char* datadir, bool textonly);
 
  protected:
   virtual bool BeginDocumentHandler();
@@ -196,20 +196,20 @@
  private:
   // We don't want to have every image in memory at once,
   // so we store some metadata as we go along producing
-  // PDFs one page at a time. At the end that metadata is
+  // PDFs one page at a time. At the end, that metadata is
   // used to make everything that isn't easily handled in a
   // streaming fashion.
   long int obj_;                     // counter for PDF objects
   GenericVector<long int> offsets_;  // offset of every PDF object in bytes
   GenericVector<long int> pages_;    // object number for every /Page object
   const char *datadir_;              // where to find the custom font
+  bool textonly_;                    // skip images if set
   // Bookkeeping only. DIY = Do It Yourself.
   void AppendPDFObjectDIY(size_t objectsize);
   // Bookkeeping + emit data.
   void AppendPDFObject(const char *data);
   // Create the /Contents object for an entire page.
-  static char* GetPDFTextObjects(TessBaseAPI* api,
-                                 double width, double height);
+  char* GetPDFTextObjects(TessBaseAPI* api, double width, double height);
   // Turn an image into a PDF object. Only transcode if we have to.
   static bool imageToPDFObj(Pix *pix, char *filename, long int objnum,
                           char **pdf_object, long int *pdf_object_size);

--- api/tesseractmain.cpp	2016-12-15 15:28:37.000000000 -0800
+++ api/tesseractmain.cpp	2017-01-19 14:50:56.000000000 -0800
@@ -337,8 +337,10 @@
 
     api->GetBoolVariable("tessedit_create_pdf", &b);
     if (b) {
-      renderers->push_back(
-          new tesseract::TessPDFRenderer(outputbase, api->GetDatapath()));
+      bool textonly;
+      api->GetBoolVariable("textonly_pdf", &textonly);
+      renderers->push_back(new tesseract::TessPDFRenderer(
+          outputbase, api->GetDatapath(), textonly));
     }
 
     api->GetBoolVariable("tessedit_write_unlv", &b);

--- ccmain/tesseractclass.cpp	2017-01-19 11:57:09.000000000 -0800
+++ ccmain/tesseractclass.cpp	2017-01-19 18:15:57.000000000 -0800
@@ -391,6 +391,8 @@
                   this->params()),
       BOOL_MEMBER(tessedit_create_pdf, false, "Write .pdf output file",
                   this->params()),
+      BOOL_MEMBER(textonly_pdf, false, "Invisible text only for PDF",
+                  this->params()),
       STRING_MEMBER(unrecognised_char, "|",
                     "Output char for unidentified blobs", this->params()),
       INT_MEMBER(suspect_level, 99, "Suspect marker level", this->params()),

--- ccmain/tesseractclass.h	2017-01-19 11:57:09.000000000 -0800
+++ ccmain/tesseractclass.h	2017-01-19 16:31:04.000000000 -0800
@@ -1027,6 +1027,7 @@
   BOOL_VAR_H(tessedit_create_hocr, false, "Write .html hOCR output file");
   BOOL_VAR_H(tessedit_create_tsv, false, "Write .tsv output file");
   BOOL_VAR_H(tessedit_create_pdf, false, "Write .pdf output file");
+  BOOL_VAR_H(textonly_pdf, false, "Invisible text only for PDF");
   STRING_VAR_H(unrecognised_char, "|",
                "Output char for unidentified blobs");
   INT_VAR_H(suspect_level, 99, "Suspect marker level");
```

**Comment by RNCTX** [comment]
@Shreeshrii string-of-image PDFs are trivially converted to TIFFs, which can be sent through Tesseract apparently.

http://kiirani.com/2013/03/22/tesseract-pdf.html

The PDF/invisible text output you guys are implementing with works quite well for me using OSX 'Preview' but for a little jerkiness depending on scaling, I have not used any other PDF viewers yet.

This is quite a big deal, in my opinion, as it will allow those who have, for instance... legal documents containing signatures, to keep their original copies unmodified from their scanners, but modify them in a clean way into searchable documents.  Thanks for this.

**Comment by Shreeshrii** [comment]
Thanks for info on pdf to images conversion for use with tesseract. 

I usually use ghostscript for the purpose e.g.

```
gs -dNOPAUSE -dBATCH  -r300x300 -sDEVICE=tiffg4  -dFirstPage=168  -dLastPage=174 -sOutputFile=sample%03d.tif ./sample.pdf
```

I will give the other suggestions a try (including a new one suggested by zdenop in the forum- https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/tesseract-ocr/vvMldrkcuOQ/xLES3_ZoEwAJ )

@jbreiden Thanks, Jeff, for this invisible text output pdf which can be merged with the original pdf.

**Comment by jbreiden** [comment]
pdfimages from poppler-utils will do image extraction as well. And pdfium offers API calls for image extraction. I am sure there are many others. Have fun.

**Comment by Wikinaut** [comment]
@jbreiden something looks to be missing in your diff:

I created a formal patch https://github.com/Wikinaut/tesseract/commit/3609964be8e664924cb32f830e64b666079fb8c2 which could be cherry-picked , but there's a still a problem with it. Pls. can you submit your changes in capi.cpp and capi.h ?
```
$ make
...
libtool: compile:  g++ -DHAVE_CONFIG_H -I. -I.. -g -Wall -O0 -DDEBUG -DLOCALEDIR=\"/usr/local/share/locale\" -DUSE_STD_NAMESPACE -I../arch -I../lstm -I../ccutil -I../ccstruct -I../viewer -I../textord -I../dict -I../classify -I../ccmain -I../wordrec -I../cutil -I../opencl -I/usr/local/include/leptonica -pthread -I/usr/include/pango-1.0 -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/cairo -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/pixman-1 -I/usr/include/freetype2 -I/usr/include/libpng12 -g -O2 -std=c++11 -MT libtesseract_api_la-capi.lo -MD -MP -MF .deps/libtesseract_api_la-capi.Tpo -c capi.cpp  -fPIC -DPIC -o .libs/libtesseract_api_la-capi.o

capi.cpp: In function 'TessResultRenderer* TessPDFRendererCreate(const char*, const char*)':
capi.cpp:69:51: error: no matching function for call to 'tesseract::TessPDFRenderer::TessPDFRenderer(const char*&, const char*&)'
     return new TessPDFRenderer(outputbase, datadir);
                                                   ^
capi.cpp:69:51: note: candidates are:
In file included from capi.h:25:0,
                 from capi.cpp:21:
renderer.h:189:3: note: tesseract::TessPDFRenderer::TessPDFRenderer(const char*, const char*, bool)
   TessPDFRenderer(const char* outputbase, const char* datadir, bool textonly);
   ^
renderer.h:189:3: note:   candidate expects 3 arguments, 2 provided
renderer.h:185:16: note: tesseract::TessPDFRenderer::TessPDFRenderer(const tesseract::TessPDFRenderer&)
 class TESS_API TessPDFRenderer : public TessResultRenderer {
                ^
renderer.h:185:16: note:   candidate expects 1 argument, 2 provided
renderer.h:185:16: note: tesseract::TessPDFRenderer::TessPDFRenderer(tesseract::TessPDFRenderer&&)
renderer.h:185:16: note:   candidate expects 1 argument, 2 provided
capi.cpp:70:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
Makefile:641: recipe for target 'libtesseract_api_la-capi.lo' failed
make[2]: *** [libtesseract_api_la-capi.lo] Error 1
```

**Comment by amitdo** [comment]
>Ray will eventually merge this patch, but it is hard to predict when. I am posting here for anyone who is impatient or excited.

I suggest to merge this to master now. Ray can modify it later if needed.

**Comment by Wikinaut** [comment]
@amitdo currently, building fails, see https://github.com/tesseract-ocr/tesseract/issues/660#issuecomment-274035871

**Comment by zdenop** [comment]
merged to master.
@Wikinaut: try master now.

**Comment by Wikinaut** [comment]
@zdenop https://github.com/tesseract-ocr/tesseract/commit/effa5741e6ef8bcb37d68250ca39c92fae85f6a9 does not work: breaks tesseract and creates broken files. Who has tested that patch, and how ?

**Comment by amitdo** [comment]
>effa574 does not work: breaks tesseract and creates broken files. Who has tested that patch, and how ?

I had the impression that Jeff tested it. Maybe I was wrong.

**Comment by Wikinaut** [comment]
@amitdo with "it breaks" I mean, that the "normal" function of tesseract is broken, https://github.com/tesseract-ocr/tesseract/commit/effa5741e6ef8bcb37d68250ca39c92fae85f6a9 always creates a blank pdf.

**Comment by Wikinaut** [comment]
And now, if I combine my original input pdf with the created output, I appear to have two text layers. Who can help, it's getting more and more complicated. 

Let's go back to the roots:
Why not simply passing the original input image to the output, inside tesseract ?

**Comment by jbarlow83** [comment]
Tested. With effa574, `tesseract -c textonly_pdf=1` works correctly and `tesseract -c textonly_pdf=0` produces an invalid PDF.

The problem is missing a "/" in front of XObject.

**Comment by jbarlow83** [comment]
Fix in #667

**Comment by Wikinaut** [comment]
@jbarlow83 works. But when I look to capi.cpp then - I think - you have to apply the corresponding change in capi.h see
https://github.com/Wikinaut/tesseract/commit/5e8089107a4bc77c3eca51f6404c41a1854e652e#diff-1ff9fac4997a03321dc873248bcf1309

(I am not sure, whether my patch is correct.))

**Comment by jbreiden** [comment]
I clearly did not test well enough to find the embarrassing /XObject bug. Thank you for finding and fixing that @jbarlow83. Regarding capi.h and capi.cpp, the development branch that I am sharing with @theraysmith  doesn't have those files. I don't know what the story is with that, maybe Ray does. @Wikinaut, there are problems with "send one image through OCR and insert a different image in the output PDF". Image management becomes difficult, especially considering TIFF and PDF are multipage formats.  On top of that, certain image formats require transcoding. I think the textonly_pdf approach makes a building block that plays well with other tools, and is the right way to go. It is super simple to implement, and is especially well suited for turning scanned pdf and turning that into to searchable pdf. That's something that increases in importance every single day.

**Comment by Wikinaut** [comment]
@jbreiden regarding "image_passthrough", pls.allow me to explain my workflow, which in my view, is quite common.

* I already _have_ mixed-mode multi-page PDFs ("input.pdf") -- for example, ocr-ed with the old tesseract.
* I already found that the new LSTM mode is very much better and want to regenerate the text layer for all my archived PDFs...
* ...without loosing the high image quality of my existing scans.

With the new `textonly_pdf`mode I managed this, but it requires this additional ugly step (marked with (*)

* split input.pdf into single pages for tesseract (use pdftoppm, or gs or whatelse)
* for each $image do `tesseract input-$image.ppm textonly-$image -c textonly_pdf pdf` command
* `pdftk textonly-*.pdf cat output "textonly.pdf"
* (*) remove text layer from input.pdf -> input-without-text.pdf
* `pdftk input-without-text.pdf multibackground textonly.pdf output new-mixed-mode.pdf`

So it's still very long way with your new option.

Please, perhaps you find a possible way when input image type is a single page (and losslessly coded)

* gif (not really needed)
* **png** ; or 
* **ppm**, pbm, pgm

(**preferred**)
 to pass-through such image types.

I think, it's possible.

**Comment by jbreiden** [comment]
There is a much simpler way to achieve the same results. Remember, for most image formats Tesseract will not mangle the images in any way when creating a PDF file.

1. Extract the images from the PDF file (don't render!). For this example, we'll assume jpeg.
2. `ls *.jpg | tesseract - result pdf`   

Now it is a little more complicated if you want what was described in the top level comment. Which is to OCR a different image than what ends up in the PDF file. For that, it would look like this.

1. Extract the images from the PDF file (don't render!)
2. Merge them to create an image-only PDF using `convertopdf` from `leptonica-progs`
3. Apply your favorite image processing operations
4. Generate a text-only PDF
5. Merge your image-only and text-only PDF

Bottom line, you want your images completely unmolested during this process. No format conversion, no uncompress + decompress cycles. Nothing. Hands off. And it will work for most normal starting points.

**Comment by Wikinaut** [comment]
why again *.jpg (step 2) ? Never ever use jpg with text files.
Please don't tell the mass about jpeg. Use png, ppm, or tif...

**Comment by Wikinaut** [comment]
I already developed code for this using -c textonly_pdf=1, thanks

**Comment by jbreiden** [comment]
A PDF file is a bag of images. If it is holding a bunch of JPEG images, extract them as-is. Don't convert. Don't recompress. Just empty the PDF bag and get your images out.

**Comment by Wikinaut** [comment]
Yes and no, why can't tesseract do this (pass-through the "bunch of input images") ?

**Comment by jbreiden** [comment]
Let's shift this discussion back to the forum. Please re-ask your most recent question there; I don't follow exactly what you are asking.

**Comment by Wikinaut** [comment]
Pls. elaborate your step
"Extract the images from the PDF file (don't render!). For this example, we'll assume jpeg."

**Comment by zdenop** [comment]
C-API should be fixed now. Thanks for finding this wikinaut.
@jbreiden capi.cpp and capi.h are C-API for tesseract that is used for tesseract wrappers (python etc.)
@Wikinaut as pointed by Jeff, please move back this discussion to [tesseract user forum](https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/tesseract-ocr/vvMldrkcuOQ/3cyAq4jSAQAJ).

**Comment by Jmuccigr** [comment]
Was there a final resolution to this request for putting back in the original images? @Wikinaut?

**Comment by jbreiden** [comment]
Yes. The final solution was to implement `tesseract -c textonly_pdf=1`  and merge with images outside of Tesseract.

**Comment by Jmuccigr** [comment]
Yeah, that doesn't work for me: `Could not set option: textonly_pdf=1`

I'm using version 3.05.00 installed via homebrew.

**Comment by Wikinaut** [comment]
@Jmuccigr I am definitely _not_ happy with the current implementation, and decided some months ago to stay silent and let other users come back with the issue (hoping, that my original proposal - pass-through the original input image without transcoding it - will be implement in forthcoming versions).

**Comment by amitdo** [comment]
>I'm using version 3.05.00

The `textonly_pdf` parameter is only available on the HEAD (4.00)

**Comment by Jmuccigr** [comment]
@Wikinaut, yeah, my workflow at some point involves adding OCR'ed text to an optimized PDF. Having the OCR step degrade the quality of that PDF kind of spoils it.

**Comment by Shreeshrii** [comment]
> The textonly_pdf parameter is only available on the HEAD (4.00)

@zdenop Please backport for 3.05. Thanks!

**Comment by zdenop** [comment]
done.

**Comment by Shreeshrii** [comment]
Thanks, @zdenop. Please also make a 3.05.01 release with the latest commit in 3.05 branch so that all these enhancements are easily accessible.

**Comment by Jmuccigr** [comment]
Just getting back to this now that 3.05.01 has hit homebrew and wanted to say that it seems to be working.

I've tested it out by running text-only tesseract on a 2x version of an image - which tends to give better results if the original dpi is too low - and then combining that text-only PDF with a PDF made from the original image, which keeps the file size down.

**Comment by wrznr** [comment]
From version [8.4.0](https://github.com/qpdf/qpdf/releases/tag/release-qpdf-8.4.0) on, [qpdf](https://github.com/qpdf/qpdf) has the options `--overlay/--underlay` for easy merging of image-only and text PDFs. E.g.,
```
$ qpdf image.pdf --underlay text.pdf -- image_txt.pdf
```

**Comment by amitdo** [comment]
@wrznr, thanks for the info.

**Comment by Jeankree** [comment]
> Implementation complete and under review by Ray. @jbarlow83 this is a good time to look at the samples above and make sure they meet your needs.
> 
> ```
> tesseract -c naked_pdf=true HelloWorld.png HelloWorld pdf
> ```
Hello,
I tried this today (tesseract v5.3.0.20221214) but I was not able to run it… I always have this error:
`Could not set option: naked_pdf=true`
My command:
`tesseract -c naked_pdf=true Duerer_Image.jpg Duerer_wText pdf`
Was this option disabled in version 5 ?

My first goal was to try to understand how it works and what it does exactly... (for merging image anf textonly pdf files).

Thank you!

**Comment by Jeankree** [comment]
> You didn't read the whole thread. The parameter name was changed to `textonly_pdf` .

Oh sorry! I thought it was another parameter! (I already know `textonly_pdf`)
And: of course, I did read the whole thread, and now I read it again to be sure; I did not see any mention about changing the name of this parameter. This is why I thought it could have been different and have another use, but now, I understand that `naked_pdf=true` was only a temporary name…
Thank you for the clarification!
